### ISSUE TITLE
focus: prevent stale notifications after reentrant focus changes

### ISSUE THREAD

**Issue by vanyastaff** [vanyastaff/flui#1040]
## Severity
P2 - correctness: public focus-change notifications can contradict current focus after callback reentry.

## Problem
`FocusManager::set_primary_focus` commits a focus change, invokes node listeners, then unconditionally publishes the original `(previous, new)` pair to manager listeners. A node listener can synchronously change focus again. The outer notification is then stale and arrives after the newer transition.

Audited commit: aa6674f1bab0a65011fcf1ac0675669a6d46fa5c.

## Current behavior
[focus.rs:86-110](https://github.com/vanyastaff/flui/blob/aa6674f1bab0a65011fcf1ac0675669a6d46fa5c/crates/flui-interaction/src/routing/focus.rs#L86-L110):
```rust
Self::notify_focus_nodes(previous.as_ref(), node.as_ref());
self.notify_listeners(previous, node);
```
With A initially focused:
1. Request B; primary focus becomes B.
2. B's node listener requests C; primary focus becomes C.
3. Manager listeners receive B -> C.
4. The original call resumes and publishes A -> B.
5. Final primary focus is C, but the last reported destination is B.

`FocusNode::request_focus` routes through this code. `FocusManager::unfocus` also calls `set_primary_focus`.

The related `finish_node_replacement` path already re-reads primary focus after callbacks and suppresses a superseded outer edge (`focus.rs:128-151`). Its existing `replacement_notification_reentry_does_not_emit_a_stale_outer_edge` test explicitly establishes that expectation.

## Expected architectural behavior
Focus notification ordering must have one defined contract across ordinary requests, unfocus, node replacement and closing. Reentrant callbacks must not publish an older transition after its replacement as though it were current. Either queue transitions and deliver coherent ordered events, or suppress superseded notification transactions with an explicit revision/commit check.

## Why this matters
Consumers of the exported `FocusChangeCallback` API cannot reconstruct current focus from the notifications. A subscriber that remembers the last destination ends on B while `primary_focus()` returns C. This can desynchronize application focus-dependent state. No actual screen-reader or IME failure is claimed here; the demonstrated defect is in the public notification contract itself.

## Evidence
A temporary test added to the existing `focus.rs::tests` module reuses `manager_with_nodes`:
```rust
    #[test]
    fn audit_reentrant_request_does_not_publish_stale_focus() {
        let (manager, nodes) = manager_with_nodes(3);
        nodes[0].request_focus();
        let edges = Rc::new(RefCell::new(Vec::new()));
        let recorded = Rc::clone(&edges);
        manager.add_listener(Rc::new(move |previous, current| {
            recorded.borrow_mut().push((
                previous.map(|node| node.id()),
                current.map(|node| node.id()),
            ));
        }));
        let next = Rc::clone(&nodes[2]);
        let intermediate = Rc::downgrade(&nodes[1]);
        nodes[1].add_listener(Rc::new(move || {
            if intermediate.upgrade().unwrap().has_primary_focus() {
                next.request_focus();
            }
        }));
        nodes[1].request_focus();
        assert!(nodes[2].has_primary_focus());
        assert_eq!(
            edges.borrow().as_slice(),
            &[(Some(nodes[1].id()), Some(nodes[2].id()))],
            "a superseded outer focus transition must not be published"
        );
    }


```
Executed:
```text
cargo nextest run -p flui-interaction --lib audit_reentrant_request_does_not_publish_stale_focus
1 failed, 489 filtered out
actual:   [(B, C), (A, B)]
expected: [(B, C)]
```
The expected trace uses the policy already enforced by the node-replacement regression test. An ordered-queue design may intentionally choose a different trace, but must replace the oracle explicitly and cannot retain the observed reverse chronology.

## Root cause
Synchronous mutation and observer delivery are interleaved without a transaction identity. Releasing RefCell borrows before callbacks avoids borrow panics, but does not make the retained event payload valid after callbacks mutate the manager.

There is a second boundary to design deliberately: `notify_listeners` clones its listener list and delivers one saved pair to all listeners. Reentry from a manager listener also needs an ordering policy; adding only a check before entering that loop may be insufficient.

## Reference architectures
Verified `.flutter` at tag 3.44.0. [Flutter FocusManager](https://github.com/flutter/flutter/blob/3.44.0/packages/flutter/lib/src/widgets/focus_manager.dart#L1915-L2001) records pending focus and schedules an application pass, separating requests from committed notifications. The useful principle is an explicit focus transaction boundary, not a requirement to copy Dart microtasks.

FLUI's existing node-replacement path already implements a smaller synchronous alternative. Ordinary focus changes should share a coherent mechanism with that path.

## Possible direction
Choose and document one notification model:
- Serialize focus transitions through a pending queue and deliver each committed transition in order.
- Retain synchronous mutation but give notification transactions revisions and suppress superseded delivery consistently, including reentry during manager listeners.

A pointer-equality check alone needs scrutiny for A -> B -> C -> B reentry, where the identity returns to an earlier value but the transaction changed.

## Risks
Changing callback ordering can affect consumers relying on immediate focus. Do not accidentally defer necessary focus restoration, lose focus-history updates, notify detached/tombstoned nodes, or introduce endless dispatch for callbacks that repeatedly redirect each other.

## Tests / Acceptance criteria
- [ ] The A -> B -> C reproducer satisfies the selected coherent notification contract.
- [ ] Cover reentry from both node and manager listeners, including multiple manager subscribers.
- [ ] Cover unfocus, closing, replacement, and A -> B -> C -> B.
- [ ] Preserve focus-history and attachment invariants.
- [ ] Document synchronous/deferred semantics and expose no stale destination as the latest committed state.
- [ ] Keep or explicitly replace the existing node-replacement oracle.

## Existing issue check
Searched open/closed issues for focus reentrancy, notifications and stale focus; inspected relevant tests and focus ADR references. #540 concerns EditableText capabilities; #552 concerns generational presentation addressing. Neither covers this notification ordering defect.

**Comment by vanyastaff** [comment]
## Additional executed evidence: structural reentry also breaks attachments

Consolidating here because the focus transaction contract must preserve both notification order and attachment invariants, rather than creating another reentry issue.

Two safe public-API probes fail in **debug and release**, each **2 run,0 passed,2 failed,0 ignored/filtered**. Dependencies were freshly built by the audit worker; parent independently reran both executables. This is P2 correctness, not memory unsafety.

1. **Detach in focus-gain listener:** FocusNode::request_focus keeps a `manager_binding.borrow()` alive through manager.request_focus and synchronous notifications ([focus_scope.rs715](https://github.com/vanyastaff/flui/blob/aa6674f1bab0a65011fcf1ac0675669a6d46fa5c/crates/flui-interaction/src/routing/focus_scope.rs#L715)). The listener detaches, reaching unbind_subtree's mutable borrow at1115: `RefCell already borrowed`.
2. **Reparent in focus-loss listener during detach:** remove_child checks the original relation, calls manager.unfocus (which invokes the listener), then clears the child's parent and binding without validating that the callback moved it. Observed `retained_by_new_parent=true, parent_matches=false, attached=false`. Even if the chosen policy lets the outer detach win, the new parent's child list and the node's parent/binding must agree.

Relevant caller: flui-widgets FocusState::dispose uses attachment.detach. The probes exercise the public lower-level APIs, not a full widget-unmount workflow.

The existing body correctly identifies missing transaction identity for notifications, but releasing borrows is only one part of the repair. Structural operations need a commit/revalidation policy across arbitrary callbacks. Do not blindly overwrite a newer attachment generation after unfocus returns.

### Reproducer

Compile this as an integration test with flui-interaction as a path dependency, or link against its freshly built debug/release rlib. No GUI, window, unsafe code, clocks or sleeps.

```rust
#![forbid(unsafe_code)]

use std::rc::Rc;

use flui_interaction::{FocusManager, routing::FocusNode};

#[test]
fn focus_gain_listener_can_detach_its_node() {
    let manager = FocusManager::new();
    let node = FocusNode::new();
    let attachment = Rc::new(manager.root_scope().attach_node(&node).unwrap());
    let weak_node = Rc::downgrade(&node);
    node.add_listener(Rc::new(move || {
        if weak_node.upgrade().unwrap().has_primary_focus() {
            attachment.detach();
        }
    }));
    node.request_focus();
    assert!(!node.is_attached());
}

#[test]
fn detach_must_not_corrupt_a_reentrant_reparent() {
    let manager = FocusManager::new();
    let root = manager.root_scope();
    let old_parent = FocusNode::new();
    let new_parent = FocusNode::new();
    let node = FocusNode::new();
    root.attach_node(&old_parent).unwrap();
    root.attach_node(&new_parent).unwrap();
    let attachment = Rc::new(old_parent.attach_node(&node).unwrap());
    node.request_focus();
    let attachment_for_listener = Rc::clone(&attachment);
    let new_parent_for_listener = Rc::clone(&new_parent);
    node.add_listener(Rc::new(move || {
        attachment_for_listener
            .reparent(&new_parent_for_listener)
            .unwrap();
    }));
    attachment.detach();
    let retained = new_parent
        .children()
        .iter()
        .any(|child| Rc::ptr_eq(child, &node));
    let correct_parent = node
        .parent()
        .is_some_and(|parent| Rc::ptr_eq(&parent, &new_parent));
    eprintln!(
        "retained_by_new_parent={retained}, parent_matches={correct_parent}, attached={}",
        node.is_attached()
    );
    assert_eq!(
        retained, correct_parent,
        "parent/child links must agree after detach reentry"
    );
}
```

### Acceptance additions

- A gain callback can retire its attachment without a RefCell panic.
- Reparenting during loss/detach produces reciprocal parent/child links and coherent binding/attachment generation under a documented winner policy.
- Cover nested detach, owner close and replacement, not just request_focus.
- Keep tests independent of stale-notification suppression: fixing the event payload alone must not satisfy these structural probes.
- Preserve callback delivery outside internal borrows and avoid arbitrary user destructors while ownership structures are mutably borrowed.

This is consistent with the existing issue's focus transaction direction; Flutter3.44.0's deferred application pass remains a reference mechanism, not a requirement to adopt Dart microtasks.

**Comment by vanyastaff** [comment]
Taking this. Reference check (flutter 3.44.0, `focus_manager.dart` `applyFocusChangesIfNeeded`): Flutter defers focus application to a microtask and publishes one `notifyListeners()` per applied transition, in chronological order — a `requestFocus` issued from inside a node's `_notify` becomes the *next* transition, never an out-of-order one. FLUI applies synchronously, so the faithful contract is: a request (or `unfocus`) issued while a notification is in flight is queued and drained after the outer notification completes, FIFO, with a bounded drain so a ping-pong listener cannot hang the UI thread. That gives one contract across ordinary requests, `unfocus`, node replacement and closing; the trace for this issue's scenario becomes `[(A, B), (B, C)]` (ordered, no lost edge) rather than the suppression trace `[(B, C)]`, and the replacement path's existing suppression oracle is replaced by the ordered one — recorded as a mapping decision. Test-first, PR to follow.

**Comment by vanyastaff** [comment]
Fixed by #1152 (merged as `78675c0b`). One contract for request, `unfocus`, node replacement and `close`: a change requested while a notification is in flight is queued and applied after it, FIFO, chronologically; queued targets are re-validated at apply time; the drain is bounded with a `warn!`; the in-flight marker is a Drop guard (a panicking listener leaves the manager usable and its queued requests are discarded with a `warn!`); listeners removed mid-dispatch are skipped. The issue's scenario now publishes `[(A, B), (B, C)]` with `primary_focus() == C`.

Reference accounting is in `crates/flui-interaction/docs/ARCHITECTURE.md` `## Mapping decisions` and `.rust-studio/specs/1040-ordered-focus-notifications/survey.md` (Flutter's microtask deferral and its 2017 origin, Compose's cancel-and-revert transactions, WHATWG's removed "locked for focus", GPUI's draw-time diff and its recorded loop, Masonry's pending slot, Qt's synchronous nesting).

Against the acceptance list: all bullets met except that `finish_node_replacement` entered from inside a listener would still publish nested — documented and `debug_assert`ed, unreachable today (its only caller runs in the frame phase). Leaving this open for you to close or keep for that residual.
